### PR TITLE
External Auth security improvements

### PR DESCRIPTION
**Pull request by Karolis2011**
- [x] Adds connection timeout
- [x] Adds new API for WI to get client that is authing IP.

### CODE REVIEW

**Comment by Karolis2011** [comment]
!review

File: code/modules/world_api/commands/misc.dm [hunk 1]
```diff
@@ -208,6 +208,38 @@
...
+	var/mob/abstract/unauthed/una = unauthed[queryparams["clienttoken"]]
+
+	if(!istype(una) || !una.client)
+		statuscode = 500
+		response = "Somethnig went horribly wrong."
+		return TRUE
+
+	if(!config.external_auth)
+		statuscode = 500
+		response = "External auth is disalowed."
```
Comment (skull132) [inline code/modules/world_api/commands/misc.dm]:
```suggestion
		response = "External auth is disallowed."
```

File: code/modules/world_api/commands/misc.dm [hunk 2]
```diff
@@ -208,6 +208,38 @@
...
+	if(!(queryparams["clienttoken"] in unauthed))
+		statuscode = 404
+		response = "Client with such token is not found."
+		return TRUE
+
+	var/mob/abstract/unauthed/una = unauthed[queryparams["clienttoken"]]
+
+	if(!istype(una) || !una.client)
+		statuscode = 500
+		response = "Somethnig went horribly wrong."
```
Comment (skull132) [inline code/modules/world_api/commands/misc.dm]:
```suggestion
		response = "Something went horribly wrong."
```

File: code/modules/world_api/commands/misc.dm [hunk 3]
```diff
@@ -208,6 +208,38 @@
 	response = "Client has been authenticated sucessfully."
 	una.ClientLogin(queryparams["key"])
 
+// Authenticates client from external system
+/datum/topic_command/get_auth_client_ip
+	name = "get_auth_client_ip"
+	description = "Authenticates client from external system."
```
Comment (skull132) [inline code/modules/world_api/commands/misc.dm]:
```suggestion
	description = "Returns the IP of the client awaiting authentication, identified by the client token."
```

File: code/modules/mob/abstract/unauthed/login.dm [hunk 4]
```diff
@@ -21,10 +22,17 @@
 	if(!config.guests_allowed && config.webint_url && config.external_auth)
 		src.OpenForumAuthWindow()
 	show_browser(src, uihtml, "window=auth;size=300x300;border=0;can_close=0;can_resize=0;can_minimize=0;titlebar=1")
+	timeout_timer = addtimer(CALLBACK(src, .proc/timeout), 900, TIMER_STOPPABLE)
+
+/mob/abstract/unauthed/proc/timeout()
+	to_chat(c, "Your login time has expired. Please relog and try again.")
```
Comment (skull132) [inline code/modules/mob/abstract/unauthed/login.dm]:
```suggestion
	if (client)
		to_chat(client, "Your login time has expired. Please relog and try again.")
```

**Pull request merged**